src/stream/src/executor/hash_join.rs

```diff
@@ -1030,27 +1058,67 @@ impl<K: HashKey, S: StateStore, const T: JoinTypePrimitive> HashJoinExecutor<K,
                     let mut append_only_matched_row = None;
                     if let Some(mut matched_rows) = matched_rows {
                         let mut matched_rows_to_clean = vec![];
+                        let mut match_band_res = side_match.half_band_condition.is_none();
                         for (matched_row_ref, matched_row) in
                             matched_rows.values_mut(&side_match.all_data_types)
                         {
                             let mut matched_row = matched_row?;
                             // TODO(yuhao-su): We should find a better way to eval the expression
                             // without concat two rows.
                             // if there are non-equi expressions
-                            let check_join_condition = if let Some(ref mut cond) = cond {
+                            let check_join_condition = if need_check_join_condition {
                                 let new_row = Self::row_concat(
                                     &row,
                                     side_update.start_pos,
                                     &matched_row.row,
                                     side_match.start_pos,
                                 );
-
-                                cond.eval_row_infallible(&new_row, |err| {
-                                    ctx.on_compute_error(err, identity)
-                                })
-                                .await
-                                .map(|s| *s.as_bool())
-                                .unwrap_or(false)
+                                // TODO: We can use binary search to start matching with
+                                // `match_band_res==true`.
```

Comment:
IIUC, this PR changes the join state's pk encoding, but it doesn't seem to utilize it properly. In my mind, we should use the `range()` interface of the `BTreeMap` to fetch the exact data according to the band condition.

Reply:
> IIUC, this PR changes the join state's pk encoding, but it doesn't seem to utilize it properly. In my mind, we should use the `range()` interface of the `BTreeMap` to fetch the exact data according to the band condition.

fixed ugly.